### PR TITLE
Adding vcfanno: VCF annotation - for SV Webinar

### PR DESCRIPTION
**Pull request by NatPRoach**
This VCFanno wrapper is something of a WIP, in that it doesn't support lua functionality at the moment due to security concerns. Can be addressed later but until we figure out to make a lua sandbox to prevent security issues moving forward with this for now.

FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: tools/vcfanno/macros.xml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0"?>
+<macros>
+    <token name="@TOOL_VERSION@">0.3.2</token>
+    <token name="@PROFILE@">18.01</token>
```
Comment (bgruening) [inline tools/vcfanno/macros.xml:4]: I think you can use 20.01

File: tools/vcfanno/macros.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            <requirement type="package">tabix</requirement>
+            <requirement type="package">samtools</requirement>
+        </requirements>
+    </xml>
+    <xml name="version_command">
+        <version_command>vcfanno --help</version_command>
+    </xml>
+    <xml name="citations">
+        <citations>
+            <citation type="bibtex">@online{medaka,
```
Comment (bgruening) [inline tools/vcfanno/macros.xml:17]: medaka?

File: tools/vcfanno/macros.xml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    <token name="@WID@"><![CDATA[
+*vcfanno* allows you to quickly annotate your VCF with any number of INFO fields from any number of VCFs or BED files. It uses a simple conf file to allow the user to specify the source annotation files and fields and how they will be added to the info of the query VCF.
+
+ - For VCF, values are pulled by name from the INFO field with special-cases of ID and FILTER to pull from those VCF columns.
+ 
+ - For BED, values are pulled from (1-based) column number.
+ 
+ - For BAM, depth (count), "mapq" and "seq" are currently supported.
+
+vcfanno is written in go.
```
Comment (bgruening) [inline tools/vcfanno/macros.xml:60]: I guess this can go.

File: tools/vcfanno/vcfanno.xml [hunk 4]
```diff
@@ -0,0 +1,264 @@
+<?xml version="1.0"?>
+<tool id="vcfanno" name="vcfanno" version="@TOOL_VERSION@+galaxy0" profile="@PROFILE@">
+    <description>VCF annotation</description>
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:3]: can we add vcfanno as bio.tools ID: https://bio.tools/vcfanno

File: tools/vcfanno/vcfanno.xml [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    <expand macro="requirements"/>
+    <expand macro="version_command"/>
+
+    <!--
+    In future versions - figure out how to safely add in lua support without posing security risk
+    (Should be doable but requires writing some lua unfortunately to make custom sandboxes)
+     -->
+    <command detect_errors="exit_code"><![CDATA[
+    cat $config_toml && 
+    cp $config_toml config.toml &&
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:17]: symlink? please use single-quotes

File: tools/vcfanno/vcfanno.xml [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    cp '$input_vcf' input.vcf &&
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:18]: symlink?

File: tools/vcfanno/vcfanno.xml [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+
+    #for $i, $annotation in enumerate($annotations):
+        #set $ext = str($annotation.type.type_select)
+        cp '$annotation.type.annotation' 'input_${i}.${ext}' &&
+        ### These sets of conversions etc shouldn't be necessary but the vcfanno in Galaxy looks for an index file regardless of compression for some reason and if not present fails.
+        #if $ext == 'vcf' or $ext == 'bed':
+            bgzip 'input_${i}.${ext}' &&
+            #set $ext = $ext + ".gz"
+        #end if
+        #if $ext[-3:] == '.gz':
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:28]: not sure what is more robust but you could also use `#if '.gz' in $ext:`

File: tools/vcfanno/vcfanno.xml [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+#if $ends:
+    '$ends'
+#end if
+#if $permissive_overlap:
+    '$permissive_overlap'
+#end if
+    config.toml
+    input.vcf
+    >
+    $out_vcf
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:44]: signle quote

**Review by bgruening — COMMENTED** [review]
.shed.yml file is missing here as well.

File: tools/vcfanno/macros.xml (hunk 2 above)
Comment (NatPRoach) [reply tools/vcfanno/macros.xml:17]: C/p mistake, will fix in next PR

File: tools/vcfanno/vcfanno.xml [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    In future versions - figure out how to safely add in lua support without posing security risk
+    (Should be doable but requires writing some lua unfortunately to make custom sandboxes)
+     -->
+    <command detect_errors="exit_code"><![CDATA[
+    ln -s '$config_toml' config.toml &&
+    ln -s '$input_vcf' input.vcf &&
+
+    #for $i, $annotation in enumerate($annotations):
+        #set $ext = str($annotation.type.type_select)
+        cp '$annotation.type.annotation' 'input_${i}.${ext}' &&
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:24]: should we symlink here?

File: tools/vcfanno/vcfanno.xml [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        ### These sets of conversions etc shouldn't be necessary but the vcfanno in Galaxy looks for an index file regardless of compression for some reason and if not present fails.
+        #if $ext == 'vcf' or $ext == 'bed':
+            bgzip 'input_${i}.${ext}' &&
+            #set $ext = $ext + ".gz"
+        #end if
+        #if '.gz' in $ext[-3:]:
+            tabix -p vcf 'input_${i}.${ext}' &&
+        #elif $ext == 'bam':
+            samtools index 'input_${i}.${ext}' &&
```
Comment (bgruening) [inline tools/vcfanno/vcfanno.xml:33]: `bam` files are always indexed in Galaxy. You could use something like `ln -s '${input.metadata.bam_index}' infile.bai`

**Review by bgruening — APPROVED** [review]

**Pull request merged**